nucypher/blockchain/eth/decorators.py

```diff
@@ -70,3 +72,36 @@ def wrapped(*args, **kwargs):
         raise InvalidChecksumAddress(message)
 
     return wrapped
+
+
+def validate_secret(func: Callable) -> Callable:
+    """Decorator to enforce correct upgrade/rollback secret in upgradeable contracts"""
+
+    parameter_name = 'existing_secret_plaintext'
+    log = Logger('Upgrade-Secret-Validator')
+
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+
+        # Check for the presence of a plaintext secret in this call
+        params = inspect.getcallargs(func, self, *args, **kwargs)
+        try:
+            secret = params[parameter_name]
+        except KeyError:  # No plaintext secret present in this call
+            found_params = ', '.join("'{}'".format(p) for p in params)
+            message = f"Incorrect use of decorator: '{parameter_name}' not found. " \
+                      f"Found parameters are: {found_params}"
+            log.debug(message)
+            raise TypeError(message)
+
+        # Validate
+        secret_hash = self.contract.functions.secretHash().call()
+        is_valid_secret = secret_hash == keccak_digest(secret)
+
+        if is_valid_secret:  # Yay!  \ (•◡•) /
+            return func(self, *args, **kwargs)
+        else:
+            message = f"The secret for upgrade/rollback {self.contract_name} is not {secret}."
```

Comment:
Do we actually want to include the `secret` here? What if it is close to the actual secret itself?

Reply:
That's a fair comment. I think we can remove that.

Reply:
@KPrasch: to add since @cygnusv is on vacation:
```suggestion
            message = f"The secret for upgrade/rollback {self.contract_name} is not valid."
```

Reply:
I'd do `provided ({secret})...` for visual confirmation of what was entered.

Reply:
:heavy_check_mark: